### ISSUE TITLE
Change_scene_to_* bugs after changing scene manually

### ISSUE THREAD

**Issue by Herbherth** [godotengine/godot#86452]
### Tested versions

- Reproducible in v4.2.1.stable.official [b09f793f5], v4.1.3.stable.mono.official [f06b6836a]

### System information

Godot v4.2.1.stable - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1650 (NVIDIA; 31.0.15.3713) - Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz (8 Threads)

### Issue description

Change scene through `get_tree().change_scene_to_packed` bugs after changing scene manually.

I followed the steps from [docs](https://docs.godotengine.org/en/stable/tutorials/scripting/change_scenes_manually.html) to change scene manually, and then used the method [change_scene_to_packed](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-method-change-scene-to-packed) in this new scene.

When starting the scene from second scene, it works as documentation states:

> The current scene node is immediately removed from the tree. From that point, [Node.get_tree](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-get-tree) called on the current (outgoing) scene will return null. [current_scene](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-property-current-scene) will be null, too, because the new scene is not available yet.
> 
> At the end of the frame, the formerly current scene, already removed from the tree, will be deleted (freed from memory) and then the new scene will be instantiated and added to the tree. [Node.get_tree](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-get-tree) and [current_scene](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-property-current-scene) will be back to working as usual.

But when I start from previous scene and change to that one using the manual steps, the second scene is never deleted from tree.

### Steps to reproduce

MRP attached bellow with just minimal code to show what is happening and 3 scenes. Note that if you start from Scene_2 it works as it should until you came back to scene 1. Also note from remote scene that new "scene_2" are stacked and never deleted if you keep going.

### Minimal reproduction project (MRP)

[ChangeSceneBug.zip](https://github.com/godotengine/godot/files/13755215/ChangeSceneBug.zip)

**Comment by AThousandShips** [comment]
You need to set `current_scene`, otherwise it won't know what to change, just adding a node to the scene doesn't update this variable

However be careful with this, as per:
> Warning: Setting this directly might not work as expected, and will not add or remove any nodes from the tree, consider using change_scene_to_file or change_scene_to_packed instead.

**Comment by Herbherth** [comment]
Oh, I see, I did read that but I thought as I was calling from a code inside another scene, and since the last scene was deleted, godot would assume the scene which the script is attached to would be the current scene
I still think that if current_scene is null it should get the first available scene on tree, but I also see why this would cause problems

The page about change scenes manually could have that clearer, it could be on the part where it says that we can delete the existing scene.

Thanks, not a bug then, just bad documentation (or interpretation)

I tested here changing the code from first scene to:
```
extends Control

var scene_to_load = preload("res://Scene_2.tscn").instantiate()

func _on_change_manually_pressed() -> void:
	get_tree().root.add_child(scene_to_load)
	get_tree().current_scene = scene_to_load
	queue_free()
```
And it worked perfectly

**Comment by AThousandShips** [comment]
I'm not sure this detail needs to be clarified, it's a pretty unfounded assumption, but we could add something, note that the documentation already makes it clear that adding it like this is outside the current scene (and therefore can't be assigned to `current_scene`)

But I'm not against adding this detail, though I'm not sure how to phrase it to make it clear but also not excessive

**Comment by Herbherth** [comment]
Just to give a fresher view of things, as you can imagine, I am new to godot but I did try to ask help from friends who uses it a lot more than me.

From my POV, I imagined that current_scene was one of that variables that you avoid changing values because it would always return what scene godot is running on that moment (and that it would be constantly changing). That's why I was confused that the scene was not being freed, because I thought that because I was running a code from that scene, it means that the current scene would be that were the code is from.

So, from my read of the documentation, when I read "The current scene node is immediately removed from the tree." or "[current_scene](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-property-current-scene) will be null" I did not thought that I would have to set the current scene before it.

The documentation of godot, in general, is great and I must say that I am not watching any tutorials to learn godot. All that I've learned was following the documentation and searching here and there. In last days I read a lot the documentation and, in general, it was enough to understand how everything works and I am really amazed about how detailed and good it is. But the [change scenes manually](https://docs.godotengine.org/en/stable/tutorials/scripting/change_scenes_manually.html) page does not, in any moment, says that if you delete the scene, no other current_scene will be set and that it should also be done manually. That is not something that newers like me would assumption because, as I said, I thought it was something that godot would handle itself.

I think that one of those warning boxes that the docs have here and there saying something like 

> If you manually delete the current scene instead of using [SceneTree.change_scene_to_file()](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-method-change-scene-to-file) or [SceneTree.change_scene_to_packed()](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-method-change-scene-to-packed), a new [SceneTree.current_scene](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-property-current-scene) must be set, otherwise it will remain null

should be enough.

**Comment by AThousandShips** [comment]
> But the change scenes manually page does not, in any moment, says that if you delete the scene, no other current_scene will be set and that it should also be done manually. That is not something that newers like me would assumption because, as I said, I thought it was something that godot would handle itself.

But that's an assumption is contradicted by the documentation? It says this won't happen, it says adding a new node *won't* be deleted and *will* be outside the scene

Your assumption isn't based on anything in the documentation, it's your own invention, and your assumption isn't intuitive, it's not based on anything except your own logic

As a counter example:
Should the documentation for GDScript say "you have to connect a signal, just naming a method the standard signal handled name is not going to connect this automatically"?

**Comment by AThousandShips** [comment]
Generally, the rule is:
Don't make assumptions outside of what's stated, and if you do re-evaluate them when you turn out wrong

The documentation can't cover every possible assumption made by users that aren't based on what's in the documentation 🙂

**Comment by Herbherth** [comment]
> Generally, the rule is:
> Don't make assumptions outside of what's stated,(...)

Exactly, the page about change scene manually, in any moment, even mentions that a variable current_scene exists, so why should I care?

Don't get me wrong, Godot is a game engine, so I think you also are (or someday were) a gamedevloper yourself, and you must know that sometimes when we are too much involved with something, some things seems obvious that are not. That's why it is important to hear feedback.

For you it is obvious that current_scene exists and we must care about it. For me, it was the first time I used this variable and just because you warned me to do so.

If I must dig the documentation based on some clues to find out why my code that I copy and pasted from there is not working, why have a page about changing manually?

Another big problem about it: it is not the kind of problem that people will found out as soon as run the code. This problem will only happens if you change manually and then change through the method, and find out what is happening can be really frustrating for some one just starting with Godot.

Well, at least now this thread will be searchable from Google if some one has this issue, so whatever.

Thanks for the help again.

**Comment by KoBeWi** [comment]
I've read the tutorial and I'm not sure what is it about. Like, why is hiding a scene even considered a valid approach when it completely breaks almost any logic that isn't GUI? (unless the old scene is paused)
It looks like something that should be completely rewritten; it should explain the Godot's scene system (i.e. `current_scene` and how `change_scene_to()` works) or at least link another tutorial that explains it. But then this issue should be moved to the docs repo.

In any case, looks like #85762 might be what you need. It covers most uses of manual scene changing.
